### ISSUE TITLE
Search gives up on small strings

### ISSUE THREAD

**Issue by volhovm** [alphapapa/org-rifle#67]
Hi!

I experience the following weird behaviour: sometimes, when I'm typing small words (I've figured out `helm-org-rifle` only works with words of length >2 (or is it for occur only?)) the search behaves as if it times out. Sometimes it's possible to revive/restart it, sometimes not. It seems to happen mostly with I'm typing in russian, which is weird... maybe `grep`ping unicode is more time consuming? If I use search for some time, the problem seems to partially disappear, which makes me think again there's some caching going on which speeds up search.

Anyway, here's what happens without my unnecessary hypotheses; for example:
1. I call `helm-org-rifle` and type e.g. "math".
2. Nothing happens. The cursor blinks. The window shows no results.
3. I type "e" so that I see "mathe"; this brings many results that are related to "mathematics", of course.
4. I erase "e" again. Sometimes the results page still shows the previous results. Sometimes it becomes empty. Sometimes it actually does what it does -- outputs all the necessary results.

Also:
- I don't seem to experience this with `helm-org-rifle-occur`; it sometimes waits for about 3-4 secs on my short words, but gives me output. It seems that the words that this bug occurs on take quite a lot time in `helm-org-rifle-occur`.
- I *do* have the same problems with `helm-org-ql-agenda-files`. But it seems to work slightly better: sometimes when I run `ql` instead of `rifle` on the same string it succeeds. But `ql` also fails on some strings similarly.

TLDR: sometimes I type something and don't get any results. When I re-type it, or add and remove letters, sometimes, it begins to work.

**Comment by volhovm** [comment]
An update: I wanted to try to debug this myself, but to my great surprise the issue disappeared when I tried to use the `master` version directly, by using `load-file`. I'm not sure whether this should be closed or not, since I'm not sure whether it actually solves the issue or temporarily masks it.

**Comment by volhovm** [comment]
Another update: the issue is more complicated than I thought it was... What I know:
* To replicate the bug:
  * Open a single reasonably big file so that a helm request takes some time.
  * Call `helm-org-rifle-current-buffer`; type, e.g. "a", wait 4 secs, see the results
  * Delete an empty line with `C-k`
  * Call `helm-org-rifle-current-bufferr`; type "a", observe rifle not returning any results
* This is on 27.1, an empty config and empty `.emacs.d` directory: I open emacs, install helm-rifle from melpa, and can replicate this.

From now on, I don't have any ideas how to debug it... for now at least. Why deleting a line could affect this? Very weird.

**Comment by alphapapa** [comment]
Hi,

Thanks for your patience.  I overlooked this issue until I saw the one you filed on org-ql.

It's hard to say what's causing this, but it's probably in Helm rather than in this package.  I've occasionally seen similar issues, but they seem rare, and since it's hard to reproduce, I'm not very concerned about it, to be honest.  I hope you understand.  :)

Something that might help would be to increase the minimum input size.  Searching for only one or two characters in a large file, returning many thousands of results, is not generally going to be very useful, anyway.

If you do want to dig into it further, you'll have to dig into Helm, and you'll have to be specific about versions and such.  However, IME it's hard to debug this kind of issue, because, e.g. if you use Edebug to step through instrumented code, the problem tends to disappear.  And the pseudo-asynchronous way that Helm handles results from Emacs functions like the ones in org-rifle and org-ql (rather than external processes) makes the code flow harder to follow.

Since it's not clear that this is a bug in this package, and there's likely not a lot that we can do about it, I'm going to leave this open as a discussion.  And I'll copy this comment on the org-ql issue.  Let me know if you have more findings.

Thanks.

**Comment by alphapapa** [comment]
> This seems to be fixed automatically in Emacs 28.1?

Please be more specific.

**Comment by volhovm** [comment]
I upgraded emacs (from 27 to 28.1) and the issue disappeared. Now `helm-org-ql` and the analogous rifle helm function work as expected. That is when I type just a few letters the helm actually pauses and does the search (computationally intensive apparently). While before it would abort after a trying for a second or so, and return no results.
